src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -77,28 +79,31 @@ protected Integer initialValue()
         }
     };
 
-
     public TrieMemoryIndex(IndexContext indexContext)
     {
         super(indexContext);
-        //TODO Do we need to follow a setting for this?
-        this.data = new MemtableTrie<>(BufferType.OFF_HEAP);
+        this.data = new MemtableTrie<>(TrieMemtable.BUFFER_TYPE);
         this.primaryKeysReducer = new PrimaryKeysReducer();
-        // MemoryIndex is per-core, so analyzer should be thread-safe..
-        this.analyzerFactory = indexContext.getAnalyzerFactory();
-        this.validator = indexContext.getValidator();
-        this.isLiteral = TypeUtil.isLiteral(validator);
     }
 
-    @Override
-    public long add(DecoratedKey key, Clustering clustering, ByteBuffer value)
+    public void add(DecoratedKey key,
+                    Clustering clustering,
+                    ByteBuffer value,
+                    LongConsumer onHeapAllocationsTracker,
+                    LongConsumer offHeapAllocationsTracker)
     {
-        synchronized (writeLock)
+        boolean locked = writeLock.tryLock();
+        if (!locked)
+        {
+            writeLock.lock();
+        }
```

Comment:
Why `tryLock` before `lock()`?
Wouldn't `writeLock.lock()` suffice if we are going to block the thread anyways until the lock is acquired? 

Consider just:
```java
writeLock.lock();
```

Reply:
You are correct in this. This code was copied from DSE and I can see no reason for this change. I have reverted to using a `synchronized` block for the index.